data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependantTokenProviderBase.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class CloudDependantTokenProviderBase extends TokenProviderBase {
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL =
+            "Error determining scope due to invalid Kusto Service Resource URL";
+    protected final Set<String> scopes = new HashSet<>();
+    protected CloudInfo cloudInfo = null;
```

Comment:
Just a suggestion, not mandatory - can we make this private and change initializeCloudInfo() to getCloudInfo()?
Make it more "singleton" style. This can prevent derived classes from accessing this variable when it is NULL.

Reply:
Ok, instead of having cloudInfo as a member, the init method now receives it as a parameter. 
So, it will not be null.

There is still the problem that fields that are dependent on cloudInfo initialization will still be null, but I'm not sure what we can do about this.